Revamp the filesystem collector to use throw-away ConstMetrics.

TheTincho opened the pull request:
Following up on #61, this PR rewrites a big chunk of the filesystem exporter to use ConstMetrics that are forgotten just after they are used. I have done the same changes in the freebsd version, but I have no way of testing them (in fact, I don't even know how to tell go to compile it for sanity checking), but I am confident it should work, as the changes are literally the same.

beorn7 commented:
I'll review this ASAP (as I was part of the discussion leading up to this PR). If somebody else gets to it before me, I won't complain, though... ;)

beorn7 commented on collector/filesystem_freebsd.go:
```diff
@@ -21,78 +21,65 @@ import (
 
 const (
-	filesystemSubsystem = "filesystem"
+	subsystem = "filesystem"
```
This is in package scope. So it's confusing for all the other collectors if you pick a very generic name here.

TheTincho replied:
Ahh, I thought that names starting with lowercase had only local scopes. Will fix.

beorn7 replied:
They do have local scope. Just that in Go, the file has no own scope. The local scope here is the package.

beorn7 commented on collector/filesystem_freebsd.go:
```diff
@@ -21,78 +21,65 @@ import (
+var (
+	filesystemLabelNames = []string{"device", "mountpoint", "fstype"}
+
+	sizeDesc = prometheus.NewDesc(
```
Since all collectors are in the same package, the naming here is polluting the namespace. Many collectors might need a `sizeDesc`. You could either prefix these names with `filesystem...`, or put them into the filesystemCollector struct (where you could also put the `subsystem`). Fundamentally, we should probably have each collector in their own package, or at least the more complex ones... but that's a different discussion, which we should not drag into this PR.

TheTincho replied:
now I understand why the unwieldy long names..:)

beorn7 commented on collector/filesystem_freebsd.go:
```diff
@@ -108,17 +95,40 @@ func (c *filesystemCollector) Update(ch chan<- prometheus.Metric) (err error) {
+			sizeDesc,
+			prometheus.GaugeValue,
+			float64(mnt[i].f_blocks) * float64(mnt[i].f_bsize),
+			mpd.device, mpd.mountPoint, mpd.fsType,
```
mpd is undefiend. (I guess this wouldn't compile on freebsd. But on linux, this file is not touched, so the error isn't detected.)

beorn7 commented on collector/filesystem_linux.go:
```diff
@@ -16,12 +16,15 @@ import (
-	procMounts          = "/proc/mounts"
-	filesystemSubsystem = "filesystem"
+	procMounts  = "/proc/mounts"
+	subsystem   = "filesystem"
```
See above.

beorn7 commented on collector/filesystem_linux.go:
```diff
@@ -32,69 +35,53 @@ type filesystemDetails struct {
+var (
+	filesystemLabelNames = []string{"device", "mountpoint", "fstype"}
+
+	sizeDesc = prometheus.NewDesc(
```
See above.

beorn7 commented:
General approach looks good. It's a pity that the node exporter is designed in a way that we cannot even use all the nicely defined descriptors for a proper Describe() implementation. But that's yet another general problem here...

In different news: I'm wondering if the common code could go to one file, and only the specific parts for linux and freebsd into the specific files.

TheTincho commented:
I have fixed the issues with formatting, and variable names and scope (I noticed that the mdadm collector has similar problems and should probably be fixed too). I have also split the common code into filesystem_common.go, I hope the way I did it makes sense.

PTAL.

beorn7 commented on collector/filesystem_common.go:
```diff
@@ -0,0 +1,113 @@
+// +build !nofilesystem
```
I believe we need to add another build tag:

`// +build linux freebsd`

Otherwise, `filesystem_common.go` would be attempted to compile but it would fail on anything but linux or freebsd. (`*_linux.go` and `*_freebsd.go` are implicitly only compiled on the respective OSs.)

TheTincho replied:
added

beorn7 commented:
LGTM otherwise. Thanks. :)
